src/mappings/entities/mapAsset.ts

```diff
@@ -0,0 +1,521 @@
+  const asset = await getAsset(ticker);
+  const formattedAmount = chainAmountToBigNumber(amount);
+  const ownerAmount = getHolderAmount(asset.ownerDid, asset.holders);
+  const otherHolders = asset.holders.filter((h) => h.did !== asset.ownerDid);
+  asset.holders = [
+    ...otherHolders,
+    {
+      did: asset.ownerDid,
+      amount: ownerAmount.minus(formattedAmount).toString(),
+    },
+  ];
+  asset.totalSupply = new BigNumber(asset.totalSupply)
+    .minus(formattedAmount)
+    .toString();
+  await asset.save();
+};
+
+const handleFreeze = async (params: Record<string, any>) => {
+  const { ticker } = params;
+  const asset = await getAsset(ticker);
+  asset.isFrozen = true;
+  await asset.save();
+};
+
+const handleUnfreeze = async (params: Record<string, any>) => {
+  const { ticker } = params;
+  const asset = await getAsset(ticker);
+  asset.isFrozen = false;
+  await asset.save();
+};
+
+const handleAcceptAssetOwnershipTransfer = async (
+  params: Record<string, any>,
+) => {
+  const { authId } = params;
+  const authorization = await getAuthorization(authId);
+  if (!authorization.ticker) return;
+  const asset = await getAsset(authorization.ticker);
+  asset.ownerDid = authorization.target;
+  await asset.save();
+  await Authorization.remove(authId);
+};
+// #endregion
+
+// #region ModuleIdEnum.Settlement
+const handleAddAndAffirmInstruction = async (
+  params: Record<string, any>,
+  extrinsic: any,
+) => {
+  const { legs } = params;
+  await Promise.all(
+    legs.map(async (l: any, i: number) => {
+      const settlementId = Number(extrinsic.events[0].event.data[2].toString());
+      await Settlement.create({
+        id: `${settlementId}-${i + 1}`,
+        settlementId,
+        from: l.from.did.toString(),
+        to: l.to.did.toString(),
+        ticker: l.asset.toString(),
+        amount: chainAmountToBigNumber(l.amount).toString(),
+      }).save();
+    }),
+  );
+};
+
+const handleAffirmInstruction = async (params: Record<string, any>) => {
+  const { instructionId } = params;
+  const settlementLegs = await getSettlementLegs(instructionId);
+  await Promise.all(
+    settlementLegs.map(async (leg) => {
+      const asset = await getAsset(leg.ticker);
+      const settlementAmount = new BigNumber(leg.amount);
+      const currentFromAmount = getHolderAmount(leg.from, asset.holders);
+      const currentToAmount = getHolderAmount(leg.to, asset.holders);
+      const otherHolders = asset.holders.filter(
+        (h) => ![leg.from, leg.to].includes(h.did),
+      );
+      asset.holders = [
+        ...otherHolders,
+        {
+          did: leg.from,
+          amount: currentFromAmount.minus(settlementAmount).toString(),
+        },
+        {
+          did: leg.to,
+          amount: currentToAmount.plus(settlementAmount).toString(),
+        },
+      ];
+      asset.totalTransfers = new BigNumber(asset.totalTransfers)
+        .plus(new BigNumber(1))
+        .toString();
+      await asset.save();
```

Comment:
I think the asset.save + Settlement.remove can be done in parallel with a `Promise.all()`

Reply:
I wanted to make sure the asset saves successfully before removing the settlement and losing it. Which shouldn't be an issue if we keep settlements.

Reply:
I think everything is wrapped in a SQL transaction per block by the framework so it _should_ be an all or nothing as it is.

I am not 100% sure thats the case though.

Reply:
Refactored to use existing settlements on event side.